infrastructure/gke-billing-export/billing.go

```diff
@@ -0,0 +1,345 @@
+			SelfLink  string
+		}
+		Containers []struct {
+			Name  string
+			Usage struct {
+				CPU    string
+				Memory string
+			}
+		}
+	}
+}
+
+type Usage struct {
+	Timestamp      time.Time `bigquery:"timestamp"`
+	Project        string    `bigquery:"project"`
+	Cluster        string    `bigquery:"cluster"`
+	Namespace      string    `bigquery:"namespace"`
+	ServiceAccount string    `bigquery:"serviceaccount"`
+	Pod            string    `bigquery:"pod"`
+	ReservedCPU    int64     `bigquery:"reserved_cpu"`
+	ReservedRAM    int64     `bigquery:"reserved_ram"`
+	UsedCPU        int64     `bigquery:"used_cpu"`
+	UsedRAM        int64     `bigquery:"used_ram"`
+}
+
+var (
+	configFile    = flag.String("config", "config.json", "Configuration file")
+	numGoroutines = flag.Int("goroutines", 10, "Number of parallel goroutines")
+)
+
+func createDataset(ctx context.Context, dataset *bigquery.Dataset) error {
+	err := dataset.Create(ctx, nil)
+	if err == nil {
+		log.Println("Created dataset")
+		return nil
+	}
+	if e, ok := err.(*googleapi.Error); ok {
+		if e.Code == 409 {
+			// Dataset already exists. This is fine.
+			return nil
+		}
+	}
+	return err
+}
+
+func createTable(ctx context.Context, table *bigquery.Table) error {
+	schema, err := bigquery.InferSchema(Usage{})
+	if err != nil {
+		return fmt.Errorf("unable to infer schema")
+	}
+	err = table.Create(ctx, &bigquery.TableMetadata{Schema: schema})
+	if err == nil {
+		log.Println("Created table")
+		return nil
+	}
+	if e, ok := err.(*googleapi.Error); ok {
+		if e.Code == 409 {
+			// Table already exists. This is fine.
+			return nil
+		}
+	}
+	return err
+}
+
+func main() {
+	flag.Parse()
+
+	// Load the configuration file from disk.
+	ctx := context.Background()
+	config, err := autoconfig.Load(ctx, *configFile)
+	if err != nil {
+		log.Printf("Unable to load configuration %q: %v", *configFile, err)
+	}
+	if err := config.Watch(ctx); err != nil {
+		log.Printf("Unable to watch configuration, auto reload is disabled", err)
+	}
+
+	// Create the BigQuery client.
+	bq, err := bigquery.NewClient(ctx, config.Get("bigquery.project"))
+	if err != nil {
+		log.Fatal(err)
+	}
+	dataset := bq.DatasetInProject(config.Get("bigquery.project"), config.Get("bigquery.dataset"))
+	if err := createDataset(ctx, dataset); err != nil {
+		log.Fatalf("Error creating dataset %s/%s: %v", config.Get("bigquery.project"), config.Get("bigquery.dataset"), err)
+	}
+
+	table := dataset.Table(config.Get("bigquery.table"))
+	if err := createTable(ctx, table); err != nil {
+		log.Fatalf("Error creating table %s/%s.%s: %v", config.Get("bigquery.project"), config.Get("bigquery.dataset"), config.Get("bigquery.table"), err)
+	}
+
+	clusters := getAllClusters(ctx, config)
+	config.AddValidator(func(old, new *autoconfig.Config) error {
+		clusters = getAllClusters(ctx, new)
+		return nil
+	})
+
+	// Start all background goroutines.
+	ch := make(chan *Cluster, *numGoroutines)
+	for i := 0; i < *numGoroutines; i++ {
+		go func() {
```

Comment:
This looks like a pretty large function to nest, it would help readability a lot if you could create the function with another function which accepts the channel as an argument and returns func. This would also make it clear which variables are used from the outer scope.

`func getMetricsFunc(ch chan *Cluster) func { ... }`

```
getMetrics := getMetricsFunc(ch)
go getMetrics()
```

Reply:
I've moved it into a separate function. I don't think there's a need for the function indirection though.